content/blog/guides/2020/2020-04-28-continous-integration-with-netlify-build-plugins/index.mdx

```diff
@@ -0,0 +1,82 @@
+slug: '/continuous-integration-with-netlify-build-plugins'
+desription: 'With the new Netlify build plugins it just got super simple to create a CI pipeline'
+draft: true
+---
+
+## What are you talking about?
+
+Okay, so let me get this straight: This blog is deployed on Netlify because we love Netlify ❤️. Seriously, they put just a nice focus on everything you need to be a happy developer: It’s well integrated with GitHub, you get fancy preview URLs for every single pull request, you can run your website with zero configuration and beyond that they offer a great Netlify CLI and API in case you want to hack their system. So everything is great? Almost! The thing is, I don’t work on this blog alone and so we needed a continuous integration (CI) pipeline to keep our blog running smoothly. We do boring stuff like running unit tests, linting the code or check the `yarn.lock` file for uncommitted changes. As I said, boring stuff but really important. In order to get this pipeline we configured a GitHub workflow to run all the checks for every pull request. We, of course, also run the workflow with every merge on master so we ended up disabling Netlify’s great **auto publishing** feature and used a GitHub action to deploy the blog manually within the workflow. Sure it works, but it feels messy and I was super unhappy to loose the full power of the Netlify ecosystem.
+
+## So, what are Netlify Build Plugins?
+
+It’s such a hot topic! They just released the beta version and you can enable it for your website right in the Netlify app. With that, you can basically extend the build pipeline and do whatever you want (well more or less).
+
+There are already good examples like:
+
+- Running Cypress end-to-end tests
+- Cache your Gatsby cache and improve the build time significantly
+- Automatically generate a sitemap
+
+I can’t wait to see more awesome plugins written by the open source community, but besides that, we can also use it to define our own flow. So coming back to the CI pipeline I mentioned earlier.
+
+## The pipeline is just a plugin
+
+Netlify plugins are basically just a JavaScript object defining certain hooks. So in this example I want to run some tasks before Netlify actually builds my application:
+
+```javascript
+module.exports = {
+  async onPreBuild({ utils }) {
+    try {
+      await utils.run.command('git diff --exit-code yarn.lock');
+    } catch (error) {
+      utils.build.failBuild(
+        'Please commit the latest updates in your yarn.lock file!'
+      );
+    }
+
+	try {
+		await utils.run.command('yarn lint');
+		await utils.run.command('yarn test:ci');
+    await utils.run.command('yarn build-storybook');
+	} catch (error) {
+    utils.build.failBuild('CI failed!');
+  }
+};
+```
+
+What do I do here?
+
+1. First, I check my `yarn.lock` file for uncommitted changes after installing the dependencies. That way I ensure the integrity of the dependencies.
+2. After that, we build our Storybook and then run the unit tests
```

Comment:
Isn't it the wrong order? And what about the linting step?
`> lint > test:ci > build-storybook`